### ISSUE TITLE
Exploring HTTP handler

### ISSUE THREAD

**Issue by acrm** [EkaSe/CoffeeToCode#3]
Create HTTP handler, that process requests to ```.html``` resources. Make handler to read requested file and transfer it's data in response. To see, that request transferred through handler and not simply taken from directory inject in the html body ```div```-block with some server-side dynamic information, e.g. text of http-request

#### References
[MSDN: Creating a Synchronous HTTP Handler](https://msdn.microsoft.com/en-us/library/ms228090(v=vs.100).aspx)
[MSDN: ASP.NET Application Life Cycle](https://msdn.microsoft.com/en-us/library/bb470252(v=vs.100).aspx)

**Comment by acrm** [comment]
Oops, didn't know that Asp.net Core don't have http-handlers. But, fortunately, there is such a thing as Middleware  - it's an equivalent of handlers for Core. 

[Migrating Handlers to Middlewares](https://docs.microsoft.com/en-us/aspnet/core/migration/http-modules)

**Comment by EkaSe** [comment]
@acrm It looks like writing responses is available only in asynchronous manner (Response.WriteAsync) in ASP.Net Core. Hopefully it won't cause problems..

**Comment by acrm** [comment]
Indeed. We haven't the async subject covered before, but I hope we will draw this picture right now. I suggest to start reading the [MSDN reference](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/) and asynchronously watch a few code examples:
```c#
private static void Sync()
{
    object res1 = F1();
    object res2 = F2(res1);  
    object res3 = F3(res1);
    object res4 = F4(res2, res3);
    object res5 = F5();
}

private static object F1()
{
    // Some long process
    return new object();
}
```

Here functions invokes strictly sequential: F1 finished all it's operations, then F2 starts ans so on. F5 doesn't depends on any previous results, but it will run only after all previous function completes.

```c#
private static void Callback()
{
    F1((res1) =>   
    {
        object res2 = F2(res1);
        object res3 = F3(res1);
        object res4 = F4(res2, res3);
    });
    object res5 = F5();
}

private static void F1(Action<object> callback)
{
   new Thread(() => {
        // Some long process
        callback(new object());
    })
    .Start();
}
```

Here a most straight-forward asynchronous approach, explicit using of another thread. F1 itself contains only two operations: instantiation of ```Thread``` object and it's start. All other time-consuming operations executed in other thread and might be completed long after F1 invocation is finished. Because of F2 and F3 functions, that requires completed result of F1 we need to use callback - lambda expression, that will be called after  long process is completed. In this code F5 might be called long before F2, because it's not required result of F1 long process and run right after F1, not in postponed callback.

It is possible to to write F1 in synchronous manner but it will make no sense in such use of callback:
```c#
private static void F1(Action<object> callback)
{
    // Some long process
    callback(new object());
}
```
 
There is a useful wrapper-class around callbacks and threads, called ```Task```, that make writing asynchronous much easier. Task object keep some lambda to invoke and it keep state of it's invocation: is not started, in progress, successfully finished, or finished with exception. It keeps result of successful execution and keeps thrown exception. Tasks provide instruments for cancellation of lambda execution.  Task itself gets thread for its lambda execution from [ThreadPool](https://docs.microsoft.com/en-us/dotnet/standard/threading/the-managed-thread-pool). Also ```Task``` class contains a few methods for chaining and combining tasks. 

Let's see, how Callback example can be re-written with use of Tasks:

```c#
private static void Tasks()
{
    F1Task().ContinueWith((res1Task) =>
    {
        Task<object> res2Task = F2Task(res1Task.Result);
        Task<object> res3Task = F3Task(res1Task.Result);
        Task.WhenAll(res2Task, res3Task).ContinueWith((results) =>
        {
            object res4 = F4(res2.Result, res3.Result);
        });
    });
    object res5 = F5();
}

private static Task<object> F1Task()
{
    return Task<object>.Factory.StartNew(() =>
    {
        // Some long process
        return new object();
    });
}
```

This code a little bit more verbose than previous, but here we obtain even higher level of asynchronicity: previously F3 runs synchronously after F2, because F4 requires results of both F2 and F3. Putting F3 in callback of F2 gives nothing, cause there is not independent code, that can run right after F2 invocation finished - F4 can runs only in callback of F3, which runs in callback of F2 - it's absolutely the same as synchronous invocation of F2, F3 and F4, just more complex. But if F2 and F3 rewritten in task-manner, than we could run them simultaneously, cause latter we can catch the moment, when they both complete - by method Tasks<object>.WhenAll(). This method creates new task, that will be completed only when all of given tasks are completed. And then we chain to this task F4 invocation by method ContinueWith().

Actually in this code are following sequence of operations:
1. Creation of F1 task and starting performing F1 operations
2. Chaining a callback to F1 task
3. Executing of F5 operations
4. Completing F1 operations
5. Creation of F2 task and starting performing F2 operations
6. Creation of F3 task and starting performing F3 operations
7. Creation of combining task
8. Chaining a callback to combining task
9. Completing combining task after F2 and F3 tasks completing
10. Performing F4 operations

Tasks mechanism are more powerful than raw callbacks and more provides more responsiveness that straight synchronous code, but synchronous code has more linear structure and easier to read. To simplify structure of code can be used ```async/await``` keywords:  

```c#
private static async void TasksAwaits()
{
    var res1 = await F1Async();
    Task<object> res2Task = F2Async(res1);
    Task<object> res3Task = F3Async(res1);
    var results = await Task.WhenAll(res2Task, res3Task); 
    object res4 = F4(res2.Result, res3.Result);
    object res5 = F5();
}

private static async Task<object> F1Async()
{
    return await Task<object>.Factory.StartNew(() =>
    {
        // Some long process
        return new object();
    });
}
```

The ```await``` keywords unwrap result of task execution and re-throw exception thrown in task.  Simplifying inner mechanics, statements ```var res1 = await F1Async(); ...``` will be compiled in code, equivalent to:
```c#
F1Async().ContinueWith((res1Task) =>
    {
        if(res1Task.IsFaulted) throw res1Task.Exception;
        var res1 = res1Task.Result;
        ...;
    });
```

**Comment by acrm** [comment]
Done
